gnomad/utils/constraint.py

```diff
@@ -0,0 +1,162 @@
+    additional_grouping: Optional[Tuple[str]] = (),
+    partition_hint: int = 100,
+    omit_methylation: bool = False,
+    use_table_group_by: bool = False,
+    singleton_expr: hl.expr.BooleanExpression = None,
+    max_af: Optional[float] = None,
+    freq_expr: hl.expr.ArrayExpression = None,
+    freq_meta_expr: hl.expr.ArrayExpression = None,
+) -> Union[hl.Table, Any]:
+    """
+    Count number of observed or possible variants by context, ref, alt, and optionally methylation_level.
+
+    :param ht: Input Hail Table.
+    :param count_singletons: Whether to count singletons. Defaults to False.
+    :param count_downsamplings: List of populations to use for downsampling counts. Defaults to ().
+    :param additional_grouping: Additional features to group by. i.e. exome_coverage. Defaults to ().
+    :param partition_hint: Target number of partitions for aggregation. Defaults to 100.
+    :param omit_methylation: Whether to omit 'methylation_level' from the grouping when counting variants. Defaults to False.
+    :param use_table_group_by: Whether to force grouping. Defaults to False.
+    :param singleton_expression: Expression for defining a singleton. Defaults to None.
+    :param max_af: Maximum variant AF to keep. By default, no AF cutoff is applied.
+    :param freq_expr: ArrayExpression of Structs with with 'AC' and 'AF' annotations.
+    :param freq_meta_expr: ArrayExpression of meta dictionaries corresponding to freq_expr.
+    :return: Table including 'variant_count' and downsampling counts if requested.
+    """
+    if freq_expr is None and (
+        count_downsamplings or max_af or (count_singletons and singleton_expr is None)
+    ):
+        logger.warning(
+            "freq_expr was not provided, using 'freq' as the frequency annotation."
+        )
+        freq_expr = ht.freq
+    if count_downsamplings and freq_meta_expr is None:
+        logger.warning(
+            "freq_meta_expr was not provided, using 'freq_meta' as the frequency metadata annotation."
+        )
+        freq_meta_expr = ht.freq_meta
+    if count_singletons and singleton_expr is None:
+        logger.warning(
+            "count_singletons is True and singleton_expr was not provided, using freq_expr[0].AC == 1 as the singleton expression."
+        )
+        # Slower, but more flexible (allows for downsampling agg's)
+        singleton_expr = freq_expr[0].AC == 1
+
+    grouping = hl.struct(context=ht.context, ref=ht.ref, alt=ht.alt)
+    if not omit_methylation:
+        grouping = grouping.annotate(methylation_level=ht.methylation_level)
+    for group in additional_grouping:
+        grouping = grouping.annotate(**{group: ht[group]})
+
+    agg = {
+        "variant_count": hl.agg.count_where(freq_expr.AF <= max_af)
+        if max_af
+        else hl.agg.count()
+    }
+    if count_singletons:
+        agg["singleton_count"] = hl.agg.count_where(singleton_expr)
+
+    if count_downsamplings:
```

Comment:
Just double check this line should be added. It wasn't here in another PR's change suggestion.

Reply:
It's not needed. Default is `()` so `for pop in count_downsamplings:` will just do nothing if `count_downsamplings` is empty